File: CriminalIntent/app/src/main/java/com/bignerdranch/android/criminalintent/database/CrimeTypeConverters.java
// 类型转换器;

package com.bignerdranch.android.criminalintent.database;

import androidx.room.TypeConverter;

import java.util.Date;
import java.util.UUID;

public class CrimeTypeConverters {

    @TypeConverter
    public String fromUUID(UUID uuid) {
        return uuid.toString();
    }
    @TypeConverter
    public UUID toUUID(String uuid) {
        return UUID.fromString(uuid);
    }
    @TypeConverter
    public long fromDate(Date date) {
        return date.getTime();
    }
    @TypeConverter
    public Date toDate(long millisSinceEpoch) {
        return new Date(millisSinceEpoch);
    }
}

File: CriminalIntent2/app/src/main/java/com/bignerdranch/android/criminalintent/database/CrimeTypeConverters.kt
package com.bignerdranch.android.criminalintent.database

import androidx.room.TypeConverter
import java.util.*

class CrimeTypeConverters() {

    @TypeConverter
    fun fromUUID(uuid: UUID?): String? {    // 将 uuid 转为 string 进入 database;
        return uuid?.toString();
    }
    @TypeConverter
    fun toUUID(uuid: String?): UUID? {
        return UUID.fromString(uuid);
    }
    @TypeConverter
    fun fromDate(date: Date?): Long? {
        return date?.time;
    }
    @TypeConverter
    fun toDate(millisSinceEpoch: Long?): Date? {
        // return if (millisSinceEpoch == null) {
        //     return null;
        // } else {
        //     return Date(millisSinceEpoch);
        // }
        return millisSinceEpoch?.let {
            Date(it);
        }
    }
}
